### ISSUE TITLE
Add support for different precision currencies

### ISSUE THREAD

**Issue by tmorehouse** [scurker/currency.js#29]
Not all currencies have two digits after the decimals.  Some have 4 digits after the decimal, some have 0 digits after the decimal, while the majority have 2 digits after the decimal.

It would be nice to have an option for the constructor top specify the precision either as a method, or during creation:

```
currency(2.51)  // 2.51
currency(2.51, 4)  // 2.5100, with precision option
```

or

```
currency.settings.precision(4)
```

Although this would change the "global" settings for all currency instances after `require`d.  It would be nice to set the formatting, decimal, separator, symbol and precision on a per usage basis (I have a site that needs to work in multiple currency formats in NodeJS, and altering the settings currently affects everything)

Many suggest that currencies be stored with 4 digits after the decimal in databases (along with the currency code in use), so maybe switching from 2 digits in the internal native format to 4 digits might also be an option.

**Comment by scurker** [comment]
I could see want to have individual localization settings if you're dealing with multiple currencies, for the cases I've primarily used it for there's always been a single currency.

However, with more precision you have to consider the potential issue of exceeding the max safe integer value. Since internally currency.js handles values as integers, anything beyond 2<sup>53</sup> - 1 may have some precision loss. Hopefully the values you're dealing with aren't that large, just something to consider.

One question I have; how would you handle calculations between multiple currencies? If you add two currencies together with different precisions, what should the final result be?

I'm glad to help but it may be a little while before I get a chance to tackle this.

**Comment by tmorehouse** [comment]
I wouldn't see the use of mathematics of different currencies, but I would see the use of a formatter that could provide different results on the same web page (i.e. one currency is USD, another is say JPY, which have different formatting rules.

I've have a prototype module created, based on this module.  Instead of returning the Currency object directly, it returns  a Currency Builder object, which returns a new class object given the options:

~~~~~~~~
var CurrencyFactory = require('currency-factory');
var usd = CurrencyFactory({symbol:'$', decimal:'.', separator:',', precision:2});
var jpy = CurrencyFactory({symbol:'¥', separator:',', precision:0});
var clf = CurrencyFactory({symbol:'UF', separator"'", decimal:',', precision:4});

console.log(usd(1000).add(2.25).format());
// -> "$1.002.25"
console.log(jpy(1000).add(2.25).format());
// -> "¥1,002"
console.log(clf(1000).add(2.25).format());
// -> "UF1'002,2500"
~~~~~~~~~~~

This way each currency can have its own options and be used on the same page or app without one currency's specification/options colliding with another's.

I'll be making repository once I have it tested out.  I would make a pull request, but the changes are rather drastic, and would be not backwards compatible :(

**Comment by tmorehouse** [comment]
For an example of different precision on the same currency, such as when dealing with gasoline prices in Canad (which are at the fraction of a cent):

~~~~~~~~~~
var CurrencyFactory = require('currency-factory');
var cad_gas = CurrencyFactory({symbol:'$', decimal:'.', separator:',', precision:3});
var cad = CurrencyFactory({symbol:'$', decimal:'.', separator:',', precision:2});

// 50.24 litres of gas at $1.048 per litre:
var price = cad(cad_gas(1.048).mul(53.24)).format();
// price -> $55.80
~~~~~~~~~~

Each currency class has an internal variable that holds the scale (`Math.pow(10,options.precision`).  

I have added a `this.valueOf()` method that returns a `Number(this.toString)` value that can be used for direct math / comparisons.

~~~~~~~~
cad(50) > 49.99 ? 'true' : 'false'; // -> 'true'
cad(50) + 2; //   -> 52 (as a number, not a currency)
~~~~~~~~

**Comment by scurker** [comment]
Any changes would need to be backwards compatible, otherwise it might make more sense to fork them into another library.

```
cad(50) > 49.99 ? 'true' : 'false'; // -> 'true'
cad(50.123) + 2; //   -> 52.12 (as a number, not a currency)
```

One issue with this approach is the potential [loss of precision](http://floating-point-gui.de/) with arithmetic or comparisons. There's an internal `currency.intValue` and `currency.value` that could be used to compare the values. Otherwise, it's generally safer to use the built in methods to ensure there's no precision loss.

**Comment by scurker** [comment]
@tmorehouse A long wait, but good news! I have a [branch](https://github.com/scurker/currency.js/tree/v1) where precision is working. There's a few breaking changes, which I intend to release as `v1.0.0` soon. 

`currency.settings` is now no longer global. Due to the nature of having multiple currencies running in parallel, it did not make sense to have a global magic settings for all. Unfortunately, if you depend on those global settings you will need to make a few changes before you can try out the latest.

Settings are now passed to the currency object as a second parameter, _including precision_! This allows you to define multiple currencies with whatever precision you want.

```js
const USD = value => currency(value, { symbol: '$', precision: 2 });
const JPY = value => currency(value, { symbol: '¥', precision: 0 });
const GAS = value => currency(value, { precision: 3 });
```

I'm still testing changes locally to ensure that I didn't miss anything significant, but if you have a moment feel free to check out that branch and see if it works for you.

**Comment by tmorehouse** [comment]
Excellent.

I will take a look at this in th next sort while.

**Comment by tmorehouse** [comment]
First quick look and I like what I see.

I still like the idea of having a  `.valueOf()` instance method for converting back to a regular float (coming from my old perl object overload days), ;-)

**Comment by tmorehouse** [comment]
The only issue I see is if one has, say two sets of numbers in the same currency but different precisions and trying to do math.  The current methods always look at intValue (because of how `parse` is written:

```js
  } else if (value instanceof currency) {
    v = value.intValue;
  }
```

Maybe if it compared precisions, and made the necessary adjustments, since `parse` does have access to the this.settings.

The other option might be to consider using an internal precision to store `intValue` (i.e. always at 4 decimals, etc. which would be  global `intPrecision` setting.

**Comment by scurker** [comment]
For `Number.MAX_SAFE_INTEGER`, that's being tracked via #8 - I'm still not sure how I want to support that given that there's still a few browsers that don't support it.

For the other changes, in [branch v1](https://github.com/scurker/currency.js/tree/v1) currency should now use the precision of the original currency being used. 

I do have a few concerns for `valueOf()`, it breaks a few tests and `valueOf()` generally trumps `toString()` for primitive values. I would be concerned it might be too much of a breaking change and am not sure of all the different ways to there would be to test that to ensure that was no regression. For your case, it might be best to extend it as a module.

**Comment by tmorehouse** [comment]
Cool on with the various precision math support :)

Yeah, valueOf can present some issues for a few things, but it does come in handy for comparing the value of two currency values.  Right now they compare alphabetically, so `20.00 > 1000000.00 === true`, which is not `true` numerically.`

**Comment by scurker** [comment]
It might be best to continue discussion of `valueOf()` in a new issue to help keep this one clean. I'd like to consider it, just want to consider the different scenarios that need to be tested.
